contracts/V4Router.sol

```diff
@@ -0,0 +1,186 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+import {Hooks} from "@uniswap/v4-core/src/libraries/Hooks.sol";
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {IUnlockCallback} from "@uniswap/v4-core/src/interfaces/callback/IUnlockCallback.sol";
+import {BalanceDelta} from "@uniswap/v4-core/src/types/BalanceDelta.sol";
+import {PoolKey} from "@uniswap/v4-core/src/types/PoolKey.sol";
+import {Currency, CurrencyLibrary} from "@uniswap/v4-core/src/types/Currency.sol";
+import {TickMath} from "@uniswap/v4-core/src/libraries/TickMath.sol";
+import {IHooks} from "@uniswap/v4-core/src/interfaces/IHooks.sol";
+import {PathKey} from "./libraries/PathKey.sol";
+import {IV4Router} from "./interfaces/IV4Router.sol";
+
+/// @title UniswapV4Router
+/// @notice Abstract contract that contains all internal logic needed for routing through Uniswap V4 pools
+abstract contract V4Router is IV4Router {
+    using CurrencyLibrary for Currency;
+
+    IPoolManager immutable poolManager;
+
+    constructor(IPoolManager _poolManager) {
+        poolManager = _poolManager;
+    }
+
+    function _v4Swap(SwapType swapType, bytes memory params) internal {
+        poolManager.unlock(abi.encode(SwapInfo(swapType, msg.sender, params)));
+    }
+
+    /// @inheritdoc IUnlockCallback
+    function unlockCallback(bytes calldata encodedSwapInfo) external override returns (bytes memory) {
+        if (msg.sender != address(poolManager)) revert NotPoolManager();
+
+        SwapInfo memory swapInfo = abi.decode(encodedSwapInfo, (SwapInfo));
+
+        if (swapInfo.swapType == SwapType.ExactInput) {
+            _swapExactInput(abi.decode(swapInfo.params, (IV4Router.ExactInputParams)), swapInfo.msgSender);
```

Comment:
i wonder if we should be trying to do gas-efficient abi decoding like we do in universal router given it saved so much gas? and then always using calldata and not memory if we never alter the parameters?
i could do that in like a gas-golfing PR later though to see if it works and how much it helps